File: source/com/barrybecker4/puzzle/sudoku/ui/SudokuPanel.java
// Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.sudoku.ui;

import com.barrybecker4.common.geometry.Location;
import com.barrybecker4.puzzle.sudoku.SudokuGenerator;
import com.barrybecker4.puzzle.sudoku.SudokuSolver;
import com.barrybecker4.puzzle.sudoku.model.board.Board;

import javax.swing.JPanel;
import java.awt.Graphics;

/**
 * Draws the current best solution to the puzzle in a panel.
 * The view in the model-view-controller pattern.
 *
 *  @author Barry Becker
 */
public final class SudokuPanel extends JPanel
                               implements RepaintListener {

    private SudokuRenderer renderer_;
    private UserInputListener inputListener;

    /**
     * Constructor. Pass in data for initial Sudoku problem.
     */
    SudokuPanel(int[][] initialData) {
        this(new Board(initialData));
    }

    /**
     * Constructor.
     */
    private SudokuPanel(Board b) {
        renderer_ = new SudokuRenderer(b);
        inputListener = new UserInputListener(renderer_);
        inputListener.addRepaintListener(this);
        addMouseListener(inputListener);
        addKeyListener(inputListener);
    }

    public void setBoard(Board b) {
        renderer_.setBoard(b);
    }

    public void setShowCandidates(boolean show) {
        renderer_.setShowCandidates(show);
        repaint();
    }

    /** Mark the users values as correct or not. */
    public void validatePuzzle() {
        inputListener.validateValues(getSolvedPuzzle());
        inputListener.useCorrectEntriesAsOriginal(getBoard());
        repaint();
    }

    private Board getSolvedPuzzle()  {
        SudokuSolver solver = new SudokuSolver();
        Board boardCopy = new Board(getBoard());
        solver.solvePuzzle(boardCopy);
        return boardCopy;
    }

    /**
     * reset to new puzzle with specified initial data.
     * @param initialData starting values.
     */
    public void reset(int[][] initialData) {
        renderer_.setBoard(new Board(initialData));
        repaint();
    }

    public void startSolving(SudokuSolver solver) {
        boolean solved = solver.solvePuzzle(getBoard(), this);
        showMessage(solved);
        inputListener.clear();
    }

    private void showMessage(boolean solved) {
        if ( solved )
            System.out.println( "The final solution is shown. the number of iterations was:" + getBoard().getNumIterations() );
        else
            System.out.println("This puzzle is not solvable!");
    }

    public void generateNewPuzzle(SudokuGenerator generator) {

        inputListener.clear();
        renderer_.setBoard(generator.generatePuzzleBoard());
        repaint();
    }

    public Board getBoard() {
        return renderer_.getBoard();
    }

    @Override
    public void valueEntered() {
        repaint();
    }

    @Override
    public void cellSelected(Location location) {
        repaint();
    }

    @Override
    public void requestValidation() {
        validatePuzzle();
    }

    /**
     * This renders the current state of the PuzzlePanel to the screen.
     */
    @Override
    protected void paintComponent( Graphics g ) {

        super.paintComponents( g );
        renderer_.render(g, inputListener.getUserEnteredValues(),
                         inputListener.getCurrentCellLocation(), getWidth(), getHeight());
        // without this we do not get key events.
        requestFocus();
    }
}

File: scala-source/com/barrybecker4/puzzle/sudoku/ui/SudokuPanel.scala
// Copyright by Barry G. Becker, 2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.sudoku.ui

import java.awt.Graphics
import javax.swing.JPanel
import com.barrybecker4.common.geometry.Location
import com.barrybecker4.puzzle.sudoku.model.Board
import com.barrybecker4.puzzle.sudoku.SudokuSolver
import com.barrybecker4.puzzle.sudoku.generation.{SimpleSudokuGenerator, SudokuGenerator}

/**
  * Draws the current best solution to the puzzle in a panel.
  * The view in the model-view-controller pattern.
  * @author Barry Becker
  */
final class SudokuPanel private(b: Board) extends JPanel with RepaintListener {

  private val renderer: SudokuRenderer = new SudokuRenderer(b)
  private val inputListener = new UserInputListener(renderer)
  inputListener.addRepaintListener(this)
  addMouseListener(inputListener)
  addKeyListener(inputListener)


  def this(initialData: Array[Array[Int]]) {
    this(new Board(initialData).updateFromInitialData().get)
  }

  def setBoard(b: Board): Unit = {
    renderer.board = b
  }

  def repaint(board: Board): Unit = {
    setBoard(board)
    super.repaint()
  }

  def setShowCandidates(show: Boolean): Unit = {
    renderer.setShowCandidates(show)
    repaint()
  }

  /** Mark the users values as correct or not. */
  def validatePuzzle(): Unit = {
    inputListener.validateValues(getSolvedPuzzle)
    inputListener.useCorrectEntriesAsOriginal(getBoard)
    repaint()
  }

  private def getSolvedPuzzle = {
    val solver = new SudokuSolver
    solver.solvePuzzle(getBoard).get
  }

  /** Reset to new puzzle with specified initial data.
    * @param initialData starting values.
    */
  def reset(initialData: Array[Array[Int]]): Unit = {
    renderer.board = new Board(initialData)
    repaint()
  }

  def startSolving(solver: SudokuSolver): Unit = {
    val solution = solver.solvePuzzle(getBoard)
    showMessage(solution)
    inputListener.clear()
  }

  private def showMessage(solution: Option[Board]): Unit = {
    if (solution.isDefined)
      println("The final solution is shown.")
    else println("This puzzle is not solvable!")
  }

  def generateNewPuzzle(generator: SudokuGenerator, size: Int): Unit = {
    inputListener.clear()
    renderer.board = generator.generatePuzzleBoard(size)
    repaint()
  }

  def getBoard: Board = renderer.board
  def valueEntered(): Unit = { repaint() }

  def cellSelected(location: Location): Unit = {
    repaint()
    // without this we do not get key events.
    requestFocus()
  }

  def requestValidation(): Unit = { validatePuzzle() }

  /** This renders the current state of the PuzzlePanel to the screen. */
  override protected def paintComponent(g: Graphics): Unit = {
    super.paintComponents(g)
    renderer.render(g, inputListener.getUserEnteredValues, inputListener.getCurrentCellLocation, getWidth, getHeight)
  }
}

